[feat] Include description (`descr`) in detailed listing

jack-morrison opened the pull request:
To accommodate a description string, this includes a small adjustment to the details formatting for variant # and file path, putting each on a new line.

Example output:
```
user@myhost:~/reframe-examples/tutorial$ reframe -C config/baseline_environs.py -c deps/parameterized.py -L
[ReFrame Setup]
  version:           4.9.0-dev.1
  command:           '/usr/local/share/reframe/bin/reframe -C config/baseline_environs.py -c deps/parameterized.py -L'
  launched by:       user@myhost
  working directory: '/home/user/reframe-examples/tutorial'
  settings files:    '<builtin>', 'config/baseline_environs.py'
  selected system:   'tutorialsys'
  check search path: '/home/user/reframe-examples/tutorial/deps/parameterized.py'
  stage directory:   '/home/user/reframe-examples/tutorial/stage'
  output directory:  '/home/user/reframe-examples/tutorial/output'
  log files:         '/tmp/rfm-ajxxgyr9.log'
  results database:  [off] '/home/user/.reframe/reports/results.db'

[List of matched checks]
- TestA %z=0 /7a14ae93
  variant: 0
  file: '/home/user/reframe-examples/tutorial/deps/parameterized.py'
  description: Test A
- TestA %z=1 /8258ae7a
  variant: 1
  file: '/home/user/reframe-examples/tutorial/deps/parameterized.py'
  description: Test A
- TestA %z=2 /423a76e9
  variant: 2
  file: '/home/user/reframe-examples/tutorial/deps/parameterized.py'
  description: Test A
- TestA %z=3 /e62d23e8
  variant: 3
  file: '/home/user/reframe-examples/tutorial/deps/parameterized.py'
  description: Test A
- TestA %z=4 /b1aa0bc1
  variant: 4
  file: '/home/user/reframe-examples/tutorial/deps/parameterized.py'
  description: Test A
- TestA %z=5 /536115e0
  variant: 5
  file: '/home/user/reframe-examples/tutorial/deps/parameterized.py'
  description: Test A
- TestB /cc291487
  variant: 0
  file: '/home/user/reframe-examples/tutorial/deps/parameterized.py'
  description: Test B
    ^TestA %z=9 /ca1c96ee
      variant: 9
      file: '/home/user/reframe-examples/tutorial/deps/parameterized.py'
      description: Test A
    ^TestA %z=8 /75b6718c
      variant: 8
      file: '/home/user/reframe-examples/tutorial/deps/parameterized.py'
      description: Test A
    ^TestA %z=7 /1d87616c
      variant: 7
      file: '/home/user/reframe-examples/tutorial/deps/parameterized.py'
      description: Test A
    ^TestA %z=6 /06c8e673
      variant: 6
      file: '/home/user/reframe-examples/tutorial/deps/parameterized.py'
      description: Test A
Found 11 check(s)
```


Closes #3458.

jack-morrison commented:
I'm having a bit of difficulty getting the new tests to run locally. It's not clear to me what's different about these test classes vs others in the project.

@vkarak, could you please give these a look and point out anything you see me doing incorrectly?

```
== warnings summary ==

unittests/resources/checks/descr_test.py:5
  /usr/local/share/reframe/unittests/resources/checks/descr_test.py:5: PytestCollectionWarning: cannot collect test class 'TestWithDescription' because it has a __init__ constructor (from: unittests/resources/checks/descr_test.py)
    @rfm.simple_test

unittests/resources/checks/descr_test.py:19
  /usr/local/share/reframe/unittests/resources/checks/descr_test.py:19: PytestCollectionWarning: cannot collect test class 'TestWithoutDescription' because it has a __init__ constructor (from: unittests/resources/checks/descr_test.py)
    @rfm.simple_test

unittests/resources/checks/descr_test.py:33
  /usr/local/share/reframe/unittests/resources/checks/descr_test.py:33: PytestCollectionWarning: cannot collect test class 'TestWithEmptyDescription' because it has a __init__ constructor (from: unittests/resources/checks/descr_test.py)
    @rfm.simple_test

```

vkarak commented on reframe/frontend/cli.py:
```diff
@@ -97,8 +97,13 @@ def dep_lines(u, *, prefix, depth=0, lines=None, printed=None,
+                    f'variant: {u.check.variant_num}',
+                    f'file: {location!r}'
+                ]
+                if hasattr(u.check, 'descr') and u.check.descr:
+                    details_fields.append(f'description: {u.check.descr}')
```
I would always add the description even if it's not set. I would do the following:

1. If the attribute is not present (pretty uncommon, as it would only happen if a user specifically does `descr = required` in a test), I would print `<undefined>`
2. If it's empty, I would print `<n/a>`.

Also as a style comment, we always leave a blank line after if/for/while etc. blocks.

vkarak reviewed (commented):
I think the unit tests are a bit over-engineered for such a trivial addition :-) This would lead us to maintain much more code. I would therefore suggest to simply extend the existing `test_list_with_details` by setting/unsetting etc. the `descr` field of some of the existing tests in `frontend_checks.py`.

jack-morrison commented:
After addressing comments, `description` is always included in detailed listings.

```
user@myhost:/usr/local/share/reframe/unittests/resources/checks$ RFM_CHECK_SEARCH_PATH=$(pwd) reframe -L
[ReFrame Setup]
  version:           4.9.0-dev.1
  command:           '/usr/local/share/reframe/bin/reframe -L'
  launched by:       user@myhost
  working directory: '/usr/local/share/reframe/unittests/resources/checks'
  settings files:    '<builtin>'
  selected system:   'generic'
  check search path: '/usr/local/share/reframe/unittests/resources/checks'
  stage directory:   '/usr/local/share/reframe/unittests/resources/checks/stage'
  output directory:  '/usr/local/share/reframe/unittests/resources/checks/output'
  log files:         '/tmp/rfm-dr1rxwp2.log'
  results database:  [off] '/home/user/.reframe/reports/results.db'

WARNING: frontend_checks.py: 'TestWithGenerator.x' is not copyable; not copyable attributes are not allowed inside the __init__() method or post-init hooks; consider setting them in another pipeline hook instead
WARNING: frontend_checks.py: 'TestWithFileObject.x' is not copyable; not copyable attributes are not allowed inside the __init__() method or post-init hooks; consider setting them in another pipeline hook instead
WARNING: skipping test 'EmptyInvalidTest': 'valid_systems' not defined
WARNING: skipping test file '/usr/local/share/reframe/unittests/resources/checks/hellocheck_copy.py': name conflict error: test 'HelloTest' from '/usr/local/share/reframe/unittests/resources/checks/hellocheck_copy.py' is already defined in '/usr/local/share/reframe/unittests/resources/checks/hellocheck.py' (rerun with '-v' for more information)
[List of matched checks]
- HelloMakeTest /0ab3180d
  variant: 0
  file: '/usr/local/share/reframe/unittests/resources/checks/hellocheck_make.py'
  description: Makefile test
- HelloTest /2b3e4546
  variant: 0
  file: '/usr/local/share/reframe/unittests/resources/checks/hellocheck.py'
  description: C Hello World test
- CompileOnlyHelloTest /ecfc0900
  variant: 0
  file: '/usr/local/share/reframe/unittests/resources/checks/hellocheck.py'
  description: Compile-only C Hello World test
- BadSetupCheck /2e7eb1ba
  variant: 0
  file: '/usr/local/share/reframe/unittests/resources/checks/frontend_checks.py'
  description: <no description provided>
- BadSetupCheckEarly /2e5c661e
  variant: 0
  file: '/usr/local/share/reframe/unittests/resources/checks/frontend_checks.py'
  description: Bad setup check early
- SanityFailureCheck /032a812c
  variant: 0
  file: '/usr/local/share/reframe/unittests/resources/checks/frontend_checks.py'
  description: <undefined>
- PerformanceFailureCheck /c4348ca0
  variant: 0
  file: '/usr/local/share/reframe/unittests/resources/checks/frontend_checks.py'
  description: <no description provided>
- CustomPerformanceFailureCheck /ccd1e572
  variant: 0
  file: '/usr/local/share/reframe/unittests/resources/checks/frontend_checks.py'
  description: <no description provided>
- CleanupFailTest /9d781ea6
  variant: 0
  file: '/usr/local/share/reframe/unittests/resources/checks/frontend_checks.py'
  description: <no description provided>
Found 9 check(s)

Log file(s) saved in '/tmp/rfm-dr1rxwp2.log'
```

jack-morrison commented on examples/tutorial/stream/stream_runonly.py:
```diff
@@ -11,6 +11,7 @@ class stream_test(rfm.RunOnlyRegressionTest):
     valid_systems = ['*']
     valid_prog_environs = ['*']
     executable = 'stream.x'
+    descr = 'Run the STREAM memory bandwidth benchmark.'
```
Has been removed.

jack-morrison commented on reframe/frontend/cli.py:
```diff
@@ -96,12 +96,26 @@ def dep_lines(u, *, prefix, depth=0, lines=None, printed=None,
+                if descr_value is None:
+                    descr_str = '<undefined>'
+                elif descr_value == '':
+                    descr_str = '<no description provided>'
```
I'd be okay with `<>`. @vkarak suggested `<n/a>` but "not applicable" didn't seem like the most correct to me in the case where no description is provided.

jack-morrison replied:
>  <n/a> could also mean "not available" :-) 

Fair enough 😄 

I'll adjust this!

vkarak replied:
Maybe `<none>` is less ambiguous then :-) I'm fine with either of the two though.

vkarak commented on unittests/resources/checks_unlisted/deps_simple.py:
```diff
@@ -51,3 +53,24 @@ def custom_deps(src, dst):
+    
+    @run_after('init')
+    def setup_deps(self):
+        self.depends_on('Test2')
```
Hmm, ok, we need to test also the detailed listing for test with dependencies, so it still makes sense to have this.

vkarak replied:
I updated the unit tests with the 67f5704656820db060f8ea702b949b18c51f95a0.

jack-morrison replied:
definitely cleaner - thank you!
